keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,840 @@
+
+###### 1. Admin creates BucketClass, User requests Bucket to be created
+
+The BucketClass represents a set of common properties shared by multiple buckets. It is used to specify the driver for creating Buckets, and also for configuring driver-specific parameters. More information about BucketClass is [here](#bucketclass)
+
+The BucketRequest is a request to create a Bucket. This resource is designed to specify properties that the application requires, such as the data protocol that this bucket should satisfy (i.e. s3, gcs or azure). More information about BucketRequest is [here](#bucketrequest)
 
-```yaml
-apiVersion: v1
-kind: Pod
-metadata:
-  name: my-app-pod
-  namespace: dev-ns
-spec:
-  serviceAccountName: [1]
-  containers:
-    - name: my-app
-      image: ...
-      volumeMounts:
-        - mountPath: /cosi [2]
-          name: cosi-vol
-  volumes:
-    - name: cosi-vol
-      csi: [3]
-        driver: cosi.sigs.k8s.io [4]
-        volumeAttributes: [5]
-          bucketAccessRequestName: <my-bar-name>
 ```
-1. the service account may be needed depending on cloud IAM integration with kubernetes.
-1. the mount path is the directory where the app will read the credentials and endpoint.
-1. this is an inline CSI volume.
-1. the name of the cosi-node-adapter.
-1. information needed by the cosi-node-adapter to verify that the bucket has been provisioned.
+    BucketRequest - br1                                   BucketClass - bc1
+    |------------------------------|                      |--------------------------------|
+    | metadata:                    |                      | deletionPolicy: delete         |
+    |   namespace: ns1             |                      | provisioner: s3.amazonaws.com  |
+    | spec:                        |                      | protocol: s3                   |
+    |   bucketClassName: bc1       |                      | parameters:                    |
+    |                              |                      |   key: value                   |
+    |------------------------------|                      |--------------------------------|
+```
+
+###### 2. COSI creates an intermediate Bucket object
+
+The ControllerManager creates a Bucket object by copying fields over from both the BucketRequest and BucketClass. This step can only proceed if the BucketClass exists. This Bucket object is an intermediate object, with its status condition `BucketReady` to false. This indicates that the Bucket is yet to be created in the OSP.
 
-The contents of the secret generated by the provisioner is expected to have the following fields in the provider specific format:
+More information about Bucket is [here](#bucket)
+
+```
+    Bucket - br-$uuid
+    |--------------------------------------|
+    | name: br-$uuid                       |
+    | spec:                                |
+    |   bucketClassName: bc1               |
+    |   protocol: s3                       |
+    |   parameters:                        |
```

Comment:
This field is meant for params that are a part of the OSP's features, however, not a part of the S3 standard. It can also be used to specify features in S3 standard not yet supported by COSI

Reply:
I meant to use S3 as an example. 

It is still possible to refer to a typed CRD defined by the provider using a key in the params map. That is exactly why we made params a generic way to allow configuration that applies to the OSP, but not to COSI itself.

Reply:
```
params:
  cosi.kubernetes.io/aws-specific-crd: $crd-name
```

I believe CSI allows secrets to be specified the same way.

Reply:
Here is an example of config that is specific to one provider: AWS

```
parameters:
  aws-access-tier: glacier  #https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-class-intro.html
```

It might be possible to use a typed field for params. That's why we went with an opaque map

Reply:
> Any reason why we still want to support the params map instead of having a ref directly?

Other than the issue with each vendor having different configurations, another reason for a generic key-value map is to allow new vendors to quickly on-board and ship COSI drivers without having to create entirely new APIs and manage their lifecycle.

That being said, I do see the benefit of typed configuration. I want to bring it up to the entire community and hear their thought about this approach. Since this is going to take a few cycles at least, is it ok to punt this post-alpha and consider the current approach as the working solution for now?

Reply:
Yes, can you add this under some future considerations section?

Reply:
will do